[flaky test] Controller TableViewsTest

https://github.com/apache/pinot/actions/runs/6430745125/job/17462295524?pr=11672

```
2023-10-06T11:28:21.9700794Z [ERROR] org.apache.pinot.controller.api.TableViewsTest.setUp -- Time elapsed: 4.090 s <<< FAILURE!
2023-10-06T11:28:21.9707427Z java.io.IOException: org.apache.pinot.common.exception.HttpErrorStatusException: Got error status code: 404 (Not Found) with reason: "Table not found" while sending request: http://localhost:18998/tables/viewsTable/externalview to controller: fv-az401-154.mqr3toex2t4uhkbi4zjcmyglea.bx.internal.cloudapp.net, version: Unknown
2023-10-06T11:28:21.9708783Z 	at org.apache.pinot.controller.helix.ControllerTest.sendGetRequest(ControllerTest.java:786)
2023-10-06T11:28:21.9713565Z 	at org.apache.pinot.controller.helix.ControllerTest.sendGetRequest(ControllerTest.java:776)
2023-10-06T11:28:21.9714340Z 	at org.apache.pinot.controller.api.TableViewsTest.getTableView(TableViewsTest.java:175)
2023-10-06T11:28:21.9720449Z 	at org.apache.pinot.controller.api.TableViewsTest.setUp(TableViewsTest.java:91)
2023-10-06T11:28:21.9721174Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2023-10-06T11:28:21.9726398Z 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2023-10-06T11:28:21.9727485Z 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2023-10-06T11:28:21.9728639Z 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
2023-10-06T11:28:21.9733763Z 	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:139)
2023-10-06T11:28:21.9734691Z 	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethodConsideringTimeout(MethodInvocationHelper.java:69)
2023-10-06T11:28:21.9740414Z 	at org.testng.internal.invokers.ConfigInvoker.invokeConfigurationMethod(ConfigInvoker.java:390)
2023-10-06T11:28:21.9741388Z 	at org.testng.internal.invokers.ConfigInvoker.invokeConfigurations(ConfigInvoker.java:325)
2023-10-06T11:28:21.9746598Z 	at org.testng.internal.invokers.TestMethodWorker.invokeBeforeClassMethods(TestMethodWorker.java:180)
2023-10-06T11:28:21.9747312Z 	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:122)
2023-10-06T11:28:21.9752055Z 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
2023-10-06T11:28:21.9752614Z 	at org.testng.TestRunner.privateRun(TestRunner.java:848)
2023-10-06T11:28:21.9757958Z 	at org.testng.TestRunner.run(TestRunner.java:621)
2023-10-06T11:28:21.9758492Z 	at org.testng.SuiteRunner.runTest(SuiteRunner.java:443)
2023-10-06T11:28:21.9759682Z 	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:437)
2023-10-06T11:28:21.9763993Z 	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:397)
2023-10-06T11:28:21.9764515Z 	at org.testng.SuiteRunner.run(SuiteRunner.java:336)
2023-10-06T11:28:21.9771106Z 	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
2023-10-06T11:28:21.9771640Z 	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:95)
2023-10-06T11:28:21.9775997Z 	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1280)
2023-10-06T11:28:21.9776487Z 	at org.testng.TestNG.runSuitesLocally(TestNG.java:1200)
2023-10-06T11:28:21.9780817Z 	at org.testng.TestNG.runSuites(TestNG.java:1114)
2023-10-06T11:28:21.9781284Z 	at org.testng.TestNG.run(TestNG.java:1082)
2023-10-06T11:28:21.9784988Z 	at org.apache.maven.surefire.testng.TestNGExecutor.run(TestNGExecutor.java:308)
2023-10-06T11:28:21.9785779Z 	at org.apache.maven.surefire.testng.TestNGXmlTestSuite.execute(TestNGXmlTestSuite.java:71)
2023-10-06T11:28:21.9790369Z 	at org.apache.maven.surefire.testng.TestNGProvider.invoke(TestNGProvider.java:113)
2023-10-06T11:28:21.9791107Z 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
2023-10-06T11:28:21.9792002Z 	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
2023-10-06T11:28:21.9796972Z 	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
2023-10-06T11:28:21.9797632Z 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
2023-10-06T11:28:21.9802416Z Caused by: org.apache.pinot.common.exception.HttpErrorStatusException: Got error status code: 404 (Not Found) with reason: "Table not found" while sending request: http://localhost:18998/tables/viewsTable/externalview to controller: fv-az401-154.mqr3toex2t4uhkbi4zjcmyglea.bx.internal.cloudapp.net, version: Unknown
2023-10-06T11:28:21.9803601Z 	at org.apache.pinot.common.utils.http.HttpClient.wrapAndThrowHttpException(HttpClient.java:448)
2023-10-06T11:28:21.9807599Z 	at org.apache.pinot.controller.helix.ControllerTest.sendGetRequest(ControllerTest.java:783)
2023-10-06T11:28:21.9808157Z 	... 33 more
2023-10-06T11:28:21.9808305Z 
```

--- a/pinot-controller/src/test/java/org/apache/pinot/controller/api/TableViewsTest.java
+++ b/pinot-controller/src/test/java/org/apache/pinot/controller/api/TableViewsTest.java
@@ -18,7 +18,6 @@
  */
 package org.apache.pinot.controller.api;
 
-import java.io.IOException;
 import java.net.HttpURLConnection;
 import java.net.URL;
 import java.util.Map;
@@ -34,12 +33,16 @@ import org.apache.pinot.spi.utils.InstanceTypeUtils;
 import org.apache.pinot.spi.utils.JsonUtils;
 import org.apache.pinot.spi.utils.builder.TableConfigBuilder;
 import org.apache.pinot.spi.utils.builder.TableNameBuilder;
+import org.apache.pinot.util.TestUtils;
 import org.testng.annotations.AfterClass;
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.DataProvider;
 import org.testng.annotations.Test;
 
-import static org.testng.Assert.*;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+import static org.testng.Assert.assertNull;
+import static org.testng.Assert.assertTrue;
 
 
 public class TableViewsTest extends ControllerTest {
@@ -67,37 +70,26 @@ public class TableViewsTest extends ControllerTest {
     tableConfig = new TableConfigBuilder(TableType.OFFLINE).setTableName(HYBRID_TABLE_NAME)
         .setNumReplicas(DEFAULT_MIN_NUM_REPLICAS).build();
     DEFAULT_INSTANCE.getHelixResourceManager().addTable(tableConfig);
-
-    // add schema for realtime table
     StreamConfig streamConfig = FakeStreamConfigUtils.getDefaultLowLevelStreamConfigs(4);
     tableConfig = new TableConfigBuilder(TableType.REALTIME).setTableName(HYBRID_TABLE_NAME)
         .setNumReplicas(DEFAULT_MIN_NUM_REPLICAS).setStreamConfigs(streamConfig.getStreamConfigsMap()).build();
     DEFAULT_INSTANCE.getHelixResourceManager().addTable(tableConfig);
 
     // Wait for external view get updated
-    long endTime = System.currentTimeMillis() + 10_000L;
-    while (System.currentTimeMillis() < endTime) {
-      Thread.sleep(100L);
-      TableViews.TableView tableView;
+    TestUtils.waitForCondition(aVoid -> {
       try {
-        tableView = getTableView(OFFLINE_TABLE_NAME, TableViews.EXTERNALVIEW, null);
-      } catch (IOException e) {
-        // Table may not be created yet.
-        continue;
+        TableViews.TableView tableView = getTableView(OFFLINE_TABLE_NAME, TableViews.EXTERNALVIEW, null);
+        if (tableView._offline == null || tableView._offline.size() != 1) {
+          return false;
+        }
+        tableView = getTableView(HYBRID_TABLE_NAME, TableViews.EXTERNALVIEW, null);
+        return tableView._offline != null && tableView._realtime != null
+            && tableView._realtime.size() == DEFAULT_NUM_SERVER_INSTANCES;
+      } catch (Exception e) {
+        // Expected before external view is created
+        return false;
       }
-      if ((tableView._offline == null) || (tableView._offline.size() != 1)) {
-        continue;
-      }
-      tableView = getTableView(HYBRID_TABLE_NAME, TableViews.EXTERNALVIEW, null);
-      if (tableView._offline == null) {
-        continue;
-      }
-      if ((tableView._realtime == null) || (tableView._realtime.size() != DEFAULT_NUM_SERVER_INSTANCES)) {
-        continue;
-      }
-      return;
-    }
-    fail("Failed to get external view updated");
+    }, 10_000L, "Failed to get external view updated");
   }
 
   @DataProvider(name = "viewProvider")

